src/main/java/bio/terra/service/snapshotbuilder/CriteriaQueryUtils.java

```diff
@@ -0,0 +1,146 @@
+    TableVariable tableVariable = TableVariable.forPrimary(tablePointer);
+
+    OccurrenceTable occurrenceTable = getOccurrenceTableFromDomainCriteria(domainCriteria);
+
+    TablePointer occurrencePointer = TablePointer.fromTableName(occurrenceTable.tableName());
+    TableVariable occurrenceVariable = TableVariable.forPrimary(occurrencePointer);
+
+    TablePointer ancestorPointer = TablePointer.fromTableName("concept_ancestor");
+    TableVariable ancestorVariable =
+        TableVariable.forJoined(
+            ancestorPointer,
+            "ancestor_concept_id",
+            new FieldVariable(
+                new FieldPointer(occurrencePointer, occurrenceTable.idColumnName()),
+                occurrenceVariable));
+
+    return new SubQueryFilterVariable(
+        new FieldVariable(new FieldPointer(tablePointer, "person_id"), tableVariable),
+        SubQueryFilterVariable.Operator.IN,
+        new Query(
+            List.of(
+                new FieldVariable(
+                    new FieldPointer(occurrencePointer, "person_id"), occurrenceVariable)),
+            List.of(occurrenceVariable, ancestorVariable),
+            new BooleanAndOrFilterVariable(
+                BooleanAndOrFilterVariable.LogicalOperator.OR,
+                List.of(
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(occurrencePointer, occurrenceTable.idColumnName()),
+                            occurrenceVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().doubleValue())),
+                    new BinaryFilterVariable(
+                        new FieldVariable(
+                            new FieldPointer(ancestorPointer, "ancestor_concept_id"),
+                            ancestorVariable),
+                        BinaryFilterVariable.BinaryOperator.EQUALS,
+                        new Literal(domainCriteria.getId().doubleValue()))))));
+  }
+
+  public static FilterVariable generateFilterForCriteria(SnapshotBuilderCriteria criteria) {
+    switch (criteria.getKind()) {
+      case LIST -> {
+        SnapshotBuilderProgramDataListCriteria listCriteria =
+            (SnapshotBuilderProgramDataListCriteria) criteria;
+        return generateFilterForListCriteria(listCriteria);
+      }
+      case RANGE -> {
+        SnapshotBuilderProgramDataRangeCriteria rangeCriteria =
+            (SnapshotBuilderProgramDataRangeCriteria) criteria;
+        return generateFilterForRangeCriteria(rangeCriteria);
+      }
+      case DOMAIN -> {
+        SnapshotBuilderDomainCriteria domainCriteria = (SnapshotBuilderDomainCriteria) criteria;
+        return generateFilterForDomainCriteria(domainCriteria);
+      }
+      default -> throw new BadRequestException(
+          String.format("Criteria kind %s not one of: domain, range, list", criteria.getKind()));
+    }
+  }
```

Comment:
A switch expression is a more concise way to write this
```suggestion
  public static FilterVariable generateFilterForCriteria(SnapshotBuilderCriteria criteria) {
    return switch (criteria.getKind()) {
      case LIST ->
        generateFilterForListCriteria((SnapshotBuilderProgramDataListCriteria) criteria);
      case RANGE ->
        generateFilterForRangeCriteria((SnapshotBuilderProgramDataRangeCriteria) criteria);
      case DOMAIN ->
        generateFilterForDomainCriteria((SnapshotBuilderDomainCriteria) criteria);
    };
```